hoomd/md/pytest/test_nlist.py

```diff
@@ -201,3 +208,355 @@ def test_logging():
+    sim.operations.computes.clear()
+    integrator = hoomd.md.Integrator(0.005)
+    lj = hoomd.md.pair.LJ(nlist, default_r_cut=1.1)
+    lj.params[('A', 'A')] = dict(epsilon=1.0, sigma=1.0)
+    integrator.forces.append(lj)
+    integrator.methods.append(hoomd.md.methods.NVE(hoomd.filter.All()))
+    sim.operations.integrator = integrator
+
+    sim.run(0)
+    pair_list = nlist.pair_list
+    if sim.device.communicator.rank == 0:
+        pair_list = set([frozenset(pair) for pair in pair_list])
+        assert pair_list == TRUE_PAIR_LIST
+
+    # modify cutoffs and check again
+    lj.r_cut[('A', 'A')] = 0.0
+    sim.run(0)
+    pair_list = nlist.pair_list
+    if sim.device.communicator.rank == 0:
+        pair_list = set([frozenset(pair) for pair in pair_list])
+        assert pair_list == TRUE_PAIR_LIST
+
+    nlist.r_cut[('A', 'A')] = 0.0
+    sim.run(0)
+    pair_list = nlist.pair_list
+    if sim.device.communicator.rank == 0:
+        pair_list = set([frozenset(pair) for pair in pair_list])
+        assert pair_list == set()
+
+    nlist.r_cut[('A', 'A')] = 1.1
+    sim.run(0)
+    pair_list = nlist.pair_list
+    if sim.device.communicator.rank == 0:
+        pair_list = set([frozenset(pair) for pair in pair_list])
+        assert pair_list == TRUE_PAIR_LIST
+
+    # NOTE: this test fails
+    # re-add and remove nlist compute
+    # sim.operations.computes.append(nlist)
+    # sim.operations.computes.clear()
+    # sim.run(0)
+    # pair_list = nlist.pair_list
+    # if sim.device.communicator.rank == 0:
+    #     pair_list = set([frozenset(pair) for pair in pair_list])
+    #     assert pair_list == set()
+
+    lj.r_cut[('A', 'A')] = 1.1
+    sim.run(0)
+    pair_list = nlist.pair_list
+    if sim.device.communicator.rank == 0:
+        pair_list = set([frozenset(pair) for pair in pair_list])
+        assert pair_list == TRUE_PAIR_LIST
+
+    # NOTE: this test fails
+    # add back nlist compute and remove force compute
+    # sim.operations.computes.append(nlist)
+    # sim.operations.integrator.forces.clear()
+    # sim.run(0)
+    # pair_list = nlist.pair_list
+    # if sim.device.communicator.rank == 0:
+    #     pair_list = set([frozenset(pair) for pair in pair_list])
+    #     assert pair_list == TRUE_PAIR_LIST
+
+
+def test_rank_local_pair_list(simulation_factory, lattice_snapshot_factory):
+
+    nlist = hoomd.md.nlist.Tree(buffer=0.0, default_r_cut=1.1)
+    sim: hoomd.Simulation = simulation_factory(lattice_snapshot_factory())
+    sim.operations.computes.append(nlist)
+
+    sim.run(0)
+
+    local_pair_list = nlist.local_pair_list
+    tag_pair_list = []
+    with sim.state.cpu_local_snapshot as data:
+        tags = data.particles.tag_with_ghost
+        for pair in local_pair_list:
+            tag_pair_list.append([tags[pair[0]], tags[pair[1]]])
+
+    set_tag_pair_list = set([frozenset(p) for p in tag_pair_list])
+    assert set_tag_pair_list.issubset(TRUE_PAIR_LIST)
+
+    if MPI4PY_IMPORTED:
```

Comment:
This below is not currently testing anything. Also, we should name the buffers something that implies what they store not just the MPI argument names.

Reply:
I was still formulating the tests here. If `mpi4py` is available I want to do a quick gather operation to check that the rank local pair list contains one copy of each in-rank pair, and a final check that the set of in-rank pairs + ghost pairs gives the true pair list.

And what buffers are you talking about here? The rank local pair list?

Reply:
`sendbuf` and `recvbuf` and `sendbuf2` and `recvbuf2`.

Reply:
Oh yes. Yes I'll give them some sensible names.